### PR TITLE
Optimize `completecases` to process only missingable columns

### PR DESCRIPTION
**Pull request by pstorozenko**
Optimization to two `completecases` methods by only processing _missingable_ columns.
Discussed in #2724

### CODE REVIEW

File: src/abstractdataframe/abstractdataframe.jl [hunk 1]
```diff
@@ -762,13 +762,20 @@ function completecases(df::AbstractDataFrame, col::Colon=:)
     end
     res = trues(size(df, 1))
     for i in 1:size(df, 2)
-        res .&= .!ismissing.(df[!, i])
+        if Missing <: eltype(df[!, i])
+            res .&= .!ismissing.(df[!, i])
+        end
     end
     res
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:769]:
```suggestion
    return res
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 2]
```diff
@@ -762,13 +762,20 @@ function completecases(df::AbstractDataFrame, col::Colon=:)
...
 end
 
-completecases(df::AbstractDataFrame, col::ColumnIndex) =
-    .!ismissing.(df[!, col])
+function completecases(df::AbstractDataFrame, col::ColumnIndex)
+    if Missing <: eltype(df[!, col])
+        .!ismissing.(df[!, col])
+    else
+        trues(size(df, 1))
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:776]:
```suggestion
        return .!ismissing.(df[!, col])
    else
        return trues(size(df, 1))
```

**Comment by bkamins** [comment]
Looks good. The comments are minor.
Can you please also post some benchmarks and double check that we properly cover all the possible cases in tests.

File: src/abstractdataframe/abstractdataframe.jl [hunk 3]
```diff
@@ -762,13 +762,20 @@ function completecases(df::AbstractDataFrame, col::Colon=:)
     end
     res = trues(size(df, 1))
     for i in 1:size(df, 2)
-        res .&= .!ismissing.(df[!, i])
+        if Missing <: eltype(df[!, i])
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:765]: It is better to extract `df[!, i]` once to a variable as it is not guaranteed that compiler will optimize this out (though it should in theory). Same comment below.

**Comment by pstorozenko** [comment]
While benchmarking, I've found a much bigger problem.
https://github.com/JuliaData/DataFrames.jl/blob/a671a3b86537a253d461ff54b87ae2c26e4e8a04/src/abstractdataframe/abstractdataframe.jl#L765
I suspect that this line is not aware of `df[!, i]` type and `.!ismissing.(df[!, i])` is slow.

----------------

Few benchamarks:
```julia
df = DataFrame(
    a = rand(1000000), b = rand(Int, 1000000),
    c = allowmissing(rand(1000000)), d = repeat([1:9; missing], 100000)
)
```

main branch timings:
```
ulia> @btime completecases(df);
  5.570 ms (32 allocations: 139.50 KiB)
julia> @btime completecases(df, :a);
  147.885 μs (9 allocations: 126.52 KiB)
julia> @btime completecases(df, :b);
  147.793 μs (9 allocations: 126.52 KiB)
julia> @btime completecases(df, :c);
  158.253 μs (9 allocations: 126.52 KiB)
julia> @btime completecases(df, :d);
  158.959 μs (9 allocations: 126.52 KiB)
```

----------------

Timings for 
```julia
res = trues(size(df, 1))
for i in 1:size(df, 2)
    v = df[!, i]
    if Missing <: eltype(v)
      res .&= .!ismissing.(v)
    end
end
```
and
```julia
function completecases(df::AbstractDataFrame, col::ColumnIndex)
    v = df[!, col]
    if Missing <: eltype(v)
        return .!ismissing.(v)
    else
        return trues(size(df, 1))
    end
end
```
```
julia> @btime completecases(df);
  3.148 ms (18 allocations: 130.88 KiB)
julia> @btime completecases(df, :a);
  12.735 μs (4 allocations: 122.25 KiB)
julia> @btime completecases(df, :b);
  12.269 μs (4 allocations: 122.25 KiB)
julia> @btime completecases(df, :c);
  158.393 μs (9 allocations: 126.52 KiB)
julia> @btime completecases(df, :d);
  159.116 μs (9 allocations: 126.52 KiB)
```

----------------

Timing for 
```julia
for i in 1:size(df, 2)
    v = df[!, i]
    if Missing <: eltype(v)
      res .&= aux(v)
    end
end
```
and 
```julia
aux(v) = .!ismissing.(v)
```

```
julia> @btime completecases(df);
  378.161 μs (16 allocations: 375.22 KiB)
```

------------------

To sum up, optimization with processing only missingable columns brought expected gain in performance.
However, more important is that `.!ismissing.(v)` is not aware of `v` type and is slow.
By making a simple barrier we gain a lot in time but some memory is allocated.
Maybe my suspicions are wrong, but there is a problem nevertheless.
Do you have any ideas on how to remove this allocations?

**Comment by bkamins** [comment]
Good catch. However, it is not related to function barrier. The core of the problem is the following I think:
```
julia> function f(x)
       r = trues(length(x))
       r .&= .!ismissing.(x)
       return r
       end
f (generic function with 1 method)

julia> function g(x)
       r = trues(length(x))
       y = .!ismissing.(x)
       r .&= y
       return r
       end
g (generic function with 1 method)

julia> x = rand([1, missing], 10^6);

julia> @benchmark f($x)
BenchmarkTools.Trial:
  memory estimate:  126.42 KiB
  allocs estimate:  4
  --------------
  minimum time:     1.214 ms (0.00% GC)
  median time:      1.247 ms (0.00% GC)
  mean time:        1.339 ms (0.19% GC)
  maximum time:     5.126 ms (75.65% GC)
  --------------
  samples:          3722
  evals/sample:     1

julia> @benchmark g($x)
BenchmarkTools.Trial:
  memory estimate:  248.66 KiB
  allocs estimate:  7
  --------------
  minimum time:     136.100 μs (0.00% GC)
  median time:      146.400 μs (0.00% GC)
  mean time:        166.911 μs (3.43% GC)
  maximum time:     8.322 ms (97.25% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

The most likely reason is that we use here `BitVector` which is not handled optimally in such loops. We could switch to `Vecror{Bool}` but I think it will be slower, and it is better to use your approach using operation splitting.

However, then it should be better to allocate an `aux` vector once and and update it in-place (so we allocate both `res` and `aux` only once).

@mbauman - the problem with broadcast fusion performance that we have here probably cannot be helped and we have to resolve it manually - right?

**Comment by pstorozenko** [comment]
As you suggested I've added and preallocated `aux` vector.
Same benchmarks as yesterday.
The first is for `res  .&= .!ismissing.(v)`, the second for `aux .= .!ismissing.(v); res .&= aux`.

```
julia> @btime completecases(df); 
  3.154 ms (18 allocations: 130.88 KiB)
  
julia> @btime completecases(df);
  375.694 μs (18 allocations: 253.00 KiB)
```

**Comment by bkamins** [comment]
Great. Looks good. Could you please add tests to make sure we cover all possible scenarios properly? Thank you!

File: src/abstractdataframe/abstractdataframe.jl [hunk 4]
```diff
@@ -761,14 +761,25 @@ function completecases(df::AbstractDataFrame, col::Colon=:)
                             "data frame with no columns"))
     end
     res = trues(size(df, 1))
+    aux = BitVector(undef, size(df, 1))
     for i in 1:size(df, 2)
-        res .&= .!ismissing.(df[!, i])
+        v = df[!, i]
+        if Missing <: eltype(v)
+            aux .= .!ismissing.(v)
+            res .&= aux
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:769]:
```suggestion
            # Disable fused broadcasting as it happens to be much slower
            aux .= .!ismissing.(v)
            res .&= aux
```

**Comment by bkamins** [comment]
@pstorozenko - do you plan to work on this? (no rush, but usually stalled PRs get outdated and it is hard to get back to them after some time). Thank you!

File: src/abstractdataframe/abstractdataframe.jl (hunk 4 above)
Comment (pstorozenko) [reply src/abstractdataframe/abstractdataframe.jl:770]: Sure, added

**Comment by pstorozenko** [comment]
Yes, but you're right, thanks for pinging. I had to think for a while, what needs testing.
I added tests for cases I could think of.

File: test/data.jl [hunk 5]
```diff
@@ -114,7 +114,14 @@ end
...
-
+    df3 = DataFrame(x = Int[1, 2, 3, 4], y = Union{Int, Missing}[1, missing, 2, 3], 
+                    z = Missing[missing, missing, missing, missing])
+
+    @test df3[completecases(df3, :x), :] ≅ df3
+    @test df3[completecases(df3, :y), :] ≅ df3[[1, 3, 4], :]
+    @test df3[completecases(df3, :z), :] ≅ df3[[], :]
+    @test df3[completecases(df3, [:y, :x]), :] ≅ df3[[1, 3, 4], :]
+    @test df3[completecases(df3, [:z, :x]), :] ≅ df3[[], :]
     @test df2[completecases(df2), :] == df2[[1, 2, 4], :]
```
Comment (bkamins) [inline test/data.jl:125]:
```suggestion
    @test completecases(df2), :] == .!ismissing.(df2.x2)
    @test @inferred(completecases(df3, :x)) == trues(nrow(df3))
    @test completecases(df3, :y) == .!ismissing.(df3.y)
    @test completecases(df3, :z) == completecases(df3, [:z, :x]) ==
          completecases(df3, [:x, :z]) == completecases(df3, [:y, :x, :z]) ==
          falses(nrow(df3))
    @test @inferred(completecases(df3, [:y, :x])) ==
          completecases(df3, [:x, :y]) == .!ismissing.(df3.y)
```

**Review by bkamins — APPROVED** [review]

**Comment by bkamins** [comment]
Looks good. I have edited the tests a bit.

**Comment by pstorozenko** [comment]
Sure thing, I tried to mimic _the design_ of the rest of tests, but a little refactor is always good.
Thanks!

**Comment by bkamins** [comment]
> I tried to mimic the design of the rest of tests

I noticed, but these tests were written very long time ago and I thought to clean them up a bit.

File: test/data.jl [hunk 6]
```diff
@@ -114,8 +114,17 @@ end
                     :auto)
     df2 = DataFrame([Union{Int, Missing}[1, 2, 3, 4], ["one", "two", missing, "four"]],
                     :auto)
-
-    @test df2[completecases(df2), :] == df2[[1, 2, 4], :]
+    df3 = DataFrame(x = Int[1, 2, 3, 4], y = Union{Int, Missing}[1, missing, 2, 3], 
+                    z = Missing[missing, missing, missing, missing])
+
+    @test completecases(df2), :] == .!ismissing.(df2.x2)
```
Comment (bkamins) [inline test/data.jl:120]:
```suggestion
    @test completecases(df2) == .!ismissing.(df2.x2)
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 7]
```diff
@@ -761,14 +761,26 @@ function completecases(df::AbstractDataFrame, col::Colon=:)
...
-    res
+    return res
 end
 
-completecases(df::AbstractDataFrame, col::ColumnIndex) =
-    .!ismissing.(df[!, col])
+function completecases(df::AbstractDataFrame, col::ColumnIndex)
+    v = df[!, col]
+    if Missing <: eltype(v)
+        return .!ismissing.(v)
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:779]:
```suggestion
        res = BitVector(undef, size(df, 1))
        res .= .!ismissing.(v)
        return res
```

**Comment by bkamins** [comment]
I pushed one additional fix that cleans inference issues.

File: src/abstractdataframe/abstractdataframe.jl [hunk 8]
```diff
@@ -761,14 +761,28 @@ function completecases(df::AbstractDataFrame, col::Colon=:)
...
 end
 
-completecases(df::AbstractDataFrame, col::ColumnIndex) =
-    .!ismissing.(df[!, col])
+function completecases(df::AbstractDataFrame, col::ColumnIndex)
+    v = df[!, col]
+    if Missing <: eltype(v)
+        res = BitVector(undef, size(df, 1))
+        res .= .!ismissing.(v)
+        return res
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:781]: Why not just this?
```suggestion
        return .!ismissing.(v)
```

Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl:781]: because it is not type stable. I have just reversed this. The current design (with `res`) has no performance penalty, but pasess `@inferred`.

Comment (nalimilan) [reply src/abstractdataframe/abstractdataframe.jl:781]: That's surprising. How about just adding `::BitVector` in the function definition? That would make it clearer that the goal is to fix inference.

Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl:781]: I thought you would ask about it :).

The `::BitVector` annotation could potentially allocate once more by converting `Vector{Bool}` to `BitVector` while what I do should guarantee only that only one allocation happens because broadcasting assignment does `copyto!` of `Broadcasted` into target.

Here is an example:
```
julia> f1(x::AbstractVector)::BitVector = .!ismissing.(x)
f1 (generic function with 1 method)

julia> f2(x::AbstractVector) = (res = BitVector(undef, length(x)); res .= .!ismissing.(x); return x)
f2 (generic function with 1 method)

julia> using SparseArrays

julia> x = sparse(1:10^7);

julia> using BenchmarkTools

julia> f1(x::AbstractVector)::BitVector = .!ismissing.(x)
f1 (generic function with 1 method)

julia> @btime f1($x);
  568.808 ms (7 allocations: 87.02 MiB)

julia> @btime f2($x);
  525.133 ms (4 allocations: 1.20 MiB)
```

File: test/data.jl [hunk 9]
```diff
@@ -128,6 +137,7 @@ end
 
     @test_throws ArgumentError completecases(DataFrame())
```
Comment (bkamins) [inline test/data.jl:138]:
```suggestion
    @test_throws ArgumentError completecases(DataFrame())
    @test_throws ArgumentError completecases(DataFrame(x=1:3), Cols())
```

**Comment by bkamins** [comment]
@nalimilan - any more comments on this?

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
@pstorozenko - please review the PR (as it was updated by me). If you can think of something to change (e.g. some more tests of corner caes) please do add them. Otherwise let me know we are done and I think we are good to merge this.

**Review by bkamins — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!
